pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,492 @@
+			defer c.consistentHashRWMutex.Unlock()
+			delete(c.consistentHashMap, eipName)
+			return nil
+		}
+		return err
+	}
+
+	// updateConsistentHash refresh the consistentHashMap
+	updateConsistentHash := func(eip *v1alpha2.ExternalIPPool) error {
+		nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if err != nil {
+			return err
+		}
+		nodes, err := c.nodeLister.List(nodeSel)
+		if err != nil {
+			return err
+		}
+		aliveNodes := c.aliveNodes()
+		// Node alive and Node labels matches with externalIPPool nodeSelector
+		var aliveAndMatchedNodes []string
+		for _, node := range nodes {
+			nodeName := node.Name
+			if aliveNodes.Has(nodeName) {
+				aliveAndMatchedNodes = append(aliveAndMatchedNodes, nodeName)
+			}
+		}
+		consistentHashMap := newNodeConsistentHashMap()
+		consistentHashMap.Add(aliveAndMatchedNodes...)
+		c.consistentHashRWMutex.Lock()
+		defer c.consistentHashRWMutex.Unlock()
+		c.consistentHashMap[eip.Name] = consistentHashMap
+		c.notify(eip.Name)
+		return nil
+	}
+
+	if err := updateConsistentHash(eip); err != nil {
+		klog.ErrorS(err, "Sync consistentHash by externalIPPool failed", "externalIPPool", eip.Name)
+		return err
+	}
+
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	node, event := nodeEvent.Node, nodeEvent.Event
+	switch event {
+	case memberlist.NodeJoin, memberlist.NodeLeave:
+		// when a Node joins cluster, all matched externalIPPools consistentHash should be updated.
+		// when a Node leave cluster, the Node may failed or deleted,
+		// if Node has been deleted, affected externalIPPool should enqueue, deleteNode handler has processed
+		// if the Node failed, externalIPPools consistentHash maybe changed, affected externalIPPool should enqueue
+		coreNode, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// means Node has been deleted, deleteNode handler has processed
+				klog.ErrorS(err, "Processing clusterNode events", "eventType", event)
+				return
+			}
+			klog.ErrorS(err, "Processing clusterNode events get node failed", "eventType", event)
+			return
+		}
+		affectedEIPs := c.filterEIPsFromNodeLabels(coreNode)
+		c.enqueueExternalIPPools(affectedEIPs)
+		klog.InfoS("Processed Node event", "eventType", event, "nodeName", node.Name, "affectedExternalIPPoolNum", len(affectedEIPs))
+	case memberlist.NodeUpdate:
+		klog.InfoS("Node event: update node", "nodeName", node.Name)
+	default:
+		klog.InfoS("Node event: unknown", "nodeName", node.Name, "event", event)
+	}
+}
+
+// aliveNodes return alive nodeName list in cluster
+func (c *Cluster) aliveNodes() sets.String {
+	aliveMembers := c.mList.Members()
+	nodes := make([]string, len(aliveMembers))
+	for i, node := range aliveMembers {
+		nodes[i] = node.Name
+	}
+	return sets.NewString(nodes...)
```

Comment:
```
nodes := sets.NewString()
for _, node := c.mList.Members() {
        nodes.Insert(node.Name)
}
return nodes
```
to avoid creating an extra slice.

Reply:
good, fixed